mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp

```diff
@@ -196,6 +196,7 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
           moduleTranslation.lookupFunction(attr.getValue());
       call = builder.CreateCall(callee, operandsRef);
       call->setCallingConv(callee->getCallingConv());
+      call->setAttributes(callee->getAttributes());
```

Comment:
Is it a requirement that call site and function attributes always match? Or is there any meaningful and legal use-case in which the two attribute sets could differ (beyond the call site only attributes)?

Reply:
AFAIK, there are some cases, e.g., calls with attributes like `always_inline`.

This is just a work-around the lack of proper `llvm.call` attributes support. When this is upstreamed, we may want to do the proper thing and make attributes first-class citizens in the dialect. For now, there is no standard way to do so.

We may want these to come from codegen, but doing so would be too much work:

- Generating in codegen (possibly cheap if reusing what we have for functions)
- `func.call -> llvm.call` conversion
- `llvm.call -> LLVM call instruction`

Two last steps are what we may wanna upstream.

Only cases in which we are being incorrect now are for the cases in which func-site and call-site may differ, but I think this is at least forward progress.

Reply:
Okay, makes sense. Maybe we should add a comment here explaining that this is not the final solution and may cause issues.

Reply:
Good point :+1: